docs/documentation/input-processing.md

```diff
@@ -0,0 +1,2016 @@
+with the examples analyzed so far: In each of the cases we have some boundary
+of the system that is crossed by primitive data, constructed by ints, floats,
+strings, dictionaries, lists and nothing (or very little) else, possibly nested
+deeply. The task then is to find out, if that data fits some constraints from
+structure or policy and transform it to an appropriate internal datatype quickly
+before it is processed further in the guts of the system.
+
+Moreover, we expect that these cases will have an object that can act as a
+representation of that boundary to the user (like `$_GET`, the form in the UI-
+Framework or the `ilDataSet` for XML-Imports) or that it is at least possible
+to create such an object. We then expect it to be possible to either completely
+internalize the validation or use a pattern to force the user to give
+constraints and transformations on the data before he may retrieve it.
+
+
+## Outlook
+
+### Improve Transformation and Validation
+
+The investigations so far suggest that the concepts of a `Transformation` and
+a `Constraint` are similar from the perspective of a data validation process,
+i.e. we need to perform a series of steps on some given primitive input to check
+constraints and derive semantically richer structures, while it is not important
+if a given step actually transforms the data or merely checks some constraint
+and passes the data on if it was successfull. This also shows in the observation
+that we indeed need to be able to incorporate checks into constructors of
+datastructure that can be processed in a way meaningful to humans.
+
+We thus propose to add a new method to the `ILIAS\Transformations\Transformation`
+interface:
+
+```php
+/**
+ * A transformation is a function from one datatype to another.
+ *
+ * It MUST NOT perform any sideeffects, i.e. it must be morally impossible to observe
+ * how often the transformation was actually performed. It MUST NOT touch the provided
+ * value, i.e. it is allowed to create new values but not to modify existing values.
+ * This would be an observable sideeffect.
+ *
+ * If you are reading this you most certainly do not want to implement this
+ * interface on your own but instead use the building blocks from the factory of
+ * the Refinery instead and stick them together like legos.
+ */
+interface Transformation {
+	/**
+	 * Perform the transformation and reify possible failures.
+	 *
+	 * If `$data->isError()`, the method MUST do nothing. It MUST transform the value
+	 * in `$data` like it would transform $data provided to `transform`. It must reify
+	 * every exception thrown in this process by returning a `Result` that `isError()`
+	 * and contains the exception that happened.
+	 *
+	 * If you simply need to implement a transformation you most probably want to
+	 * implement transform and derive this via the trait `DeriveTransformationInterface`.
+	 *
+	 * If you simply want to call the transformation, you most probably want to use
+	
```

Comment:
There is already a standard for factory interfaces that could fit this purpose: https://www.php-fig.org/psr/psr-17/

Reply:
Okay I think, I misread that :sweat_smile: nvm.